File: Algorithms/DLX/Elixir/dlx.exs
#!/usr/bin/env elixir
# Elixir Dancing Links (DLX) Sudoku Solver
# Port of OCaml/C DLX implementation using ETS for mutable node storage

defmodule DLX do
  # Node structure stored in ETS: {id, left, right, up, down, column, size, row_id, col_id}

  def main(args) do
    start_time = System.monotonic_time(:millisecond)

    args
    |> Enum.filter(&String.ends_with?(&1, ".matrix"))
    |> Enum.each(&solve_puzzle/1)

    end_time = System.monotonic_time(:millisecond)
    elapsed = (end_time - start_time) / 1000.0
    IO.puts("Seconds to process #{:erlang.float_to_binary(elapsed, decimals: 3)}")
  end

  defp solve_puzzle(filename) do
    IO.puts(filename)

    case read_matrix_file(filename) do
      {:ok, puzzle} ->
        print_raw_board(puzzle)
        print_puzzle(puzzle)

        # Create ETS table for nodes: {id, {left, right, up, down, column, size, row_id, col_id}}
        nodes = :ets.new(:dlx_nodes, [:set, :public])

        # Create iteration counter
        {:ok, counter} = Agent.start_link(fn -> 0 end)

        # Create solution array
        {:ok, solution_agent} = Agent.start_link(fn -> List.duplicate(-1, 81) end)

        # Build DLX matrix
        {root_id, row_metadata, clue_rows} = build_dlx_matrix(nodes, puzzle)

        # Cover clue rows before search
        cover_clue_rows(nodes, clue_rows)

        # Solve
        if dlx_search(nodes, root_id, 0, solution_agent, counter) do
          solution = Agent.get(solution_agent, & &1)
          solved = decode_solution(solution, puzzle, row_metadata)
          print_puzzle(solved)
          iterations = Agent.get(counter, & &1)
          IO.puts("\nSolved in Iterations=#{iterations}\n")
        else
          IO.puts("No solution found")
        end

        Agent.stop(counter)
        Agent.stop(solution_agent)
        :ets.delete(nodes)

      {:error, reason} ->
        IO.puts("Error reading file: #{reason}")
    end
  end

  defp read_matrix_file(filename) do
    case File.read(filename) do
      {:ok, content} ->
        lines =
          content
          |> String.split("\n", trim: true)
          |> Enum.reject(&String.starts_with?(&1, "#"))
          |> Enum.take(9)

        puzzle =
          lines
          |> Enum.map(fn line ->
            line
            |> String.split(~r/\s+/, trim: true)
            |> Enum.map(&String.to_integer/1)
          end)

        {:ok, puzzle}

      error ->
        error
    end
  end

  defp print_raw_board(puzzle) do
    Enum.each(puzzle, fn row ->
      IO.puts(Enum.join(row, " ") <> " ")
    end)
  end

  defp print_puzzle(puzzle) do
    IO.puts("\nPuzzle:")

    Enum.each(puzzle, fn row ->
      IO.puts(Enum.join(row, " ") <> " ")
    end)
  end

  # ETS node operations
  defp make_node(nodes, id, row_id, col_id) do
    # {left, right, up, down, column, size, row_id, col_id}
    node = {id, id, id, id, id, 0, row_id, col_id}
    :ets.insert(nodes, {id, node})
    id
  end

  defp get_node(nodes, id) do
    case :ets.lookup(nodes, id) do
      [{^id, node}] -> node
      [] -> raise "Node #{id} not found"
    end
  end

  defp update_node(nodes, id, node) do
    :ets.insert(nodes, {id, node})
  end

  defp node_left({left, _, _, _, _, _, _, _}), do: left
  defp node_right({_, right, _, _, _, _, _, _}), do: right
  defp node_up({_, _, up, _, _, _, _, _}), do: up
  defp node_down({_, _, _, down, _, _, _, _}), do: down
  defp node_column({_, _, _, _, column, _, _, _}), do: column
  defp node_size({_, _, _, _, _, size, _, _}), do: size
  defp node_row_id({_, _, _, _, _, _, row_id, _}), do: row_id

  defp set_left(node, left), do: {left, elem(node, 1), elem(node, 2), elem(node, 3), elem(node, 4), elem(node, 5), elem(node, 6), elem(node, 7)}
  defp set_right(node, right), do: {elem(node, 0), right, elem(node, 2), elem(node, 3), elem(node, 4), elem(node, 5), elem(node, 6), elem(node, 7)}
  defp set_up(node, up), do: {elem(node, 0), elem(node, 1), up, elem(node, 3), elem(node, 4), elem(node, 5), elem(node, 6), elem(node, 7)}
  defp set_down(node, down), do: {elem(node, 0), elem(node, 1), elem(node, 2), down, elem(node, 4), elem(node, 5), elem(node, 6), elem(node, 7)}
  defp set_column(node, col), do: {elem(node, 0), elem(node, 1), elem(node, 2), elem(node, 3), col, elem(node, 5), elem(node, 6), elem(node, 7)}
  defp set_size(node, size), do: {elem(node, 0), elem(node, 1), elem(node, 2), elem(node, 3), elem(node, 4), size, elem(node, 6), elem(node, 7)}

  # Cover a column in the DLX matrix
  defp cover_column(nodes, col_id) do
    col = get_node(nodes, col_id)
    right_id = node_right(col)
    left_id = node_left(col)

    # Remove column header from the header list
    right = get_node(nodes, right_id)
    left = get_node(nodes, left_id)
    update_node(nodes, right_id, set_left(right, left_id))
    update_node(nodes, left_id, set_right(left, right_id))

    # Cover all rows in this column
    cover_rows(nodes, node_down(col), col_id)
  end

  defp cover_rows(_nodes, row_id, col_id) when row_id == col_id, do: :ok

  defp cover_rows(nodes, row_id, col_id) do
    row_node = get_node(nodes, row_id)
    cover_row_nodes(nodes, node_right(row_node), row_id)
    cover_rows(nodes, node_down(row_node), col_id)
  end

  defp cover_row_nodes(_nodes, node_id, row_id) when node_id == row_id, do: :ok

  defp cover_row_nodes(nodes, node_id, row_id) do
    node = get_node(nodes, node_id)
    up_id = node_up(node)
    down_id = node_down(node)
    col_id = node_column(node)

    # Remove node from column
    down = get_node(nodes, down_id)
    up = get_node(nodes, up_id)
    col = get_node(nodes, col_id)

    update_node(nodes, down_id, set_up(down, up_id))
    update_node(nodes, up_id, set_down(up, down_id))
    update_node(nodes, col_id, set_size(col, node_size(col) - 1))

    cover_row_nodes(nodes, node_right(node), row_id)
  end

  # Uncover a column (exact reverse of cover)
  defp uncover_column(nodes, col_id) do
    col = get_node(nodes, col_id)

    # Uncover all rows in this column (in reverse order)
    uncover_rows(nodes, node_up(col), col_id)

    # Restore column header to the header list
    right_id = node_right(col)
    left_id = node_left(col)
    right = get_node(nodes, right_id)
    left = get_node(nodes, left_id)
    update_node(nodes, right_id, set_left(right, col_id))
    update_node(nodes, left_id, set_right(left, col_id))
  end

  defp uncover_rows(_nodes, row_id, col_id) when row_id == col_id, do: :ok

  defp uncover_rows(nodes, row_id, col_id) do
    row_node = get_node(nodes, row_id)
    uncover_row_nodes(nodes, node_left(row_node), row_id)
    uncover_rows(nodes, node_up(row_node), col_id)
  end

  defp uncover_row_nodes(_nodes, node_id, row_id) when node_id == row_id, do: :ok

  defp uncover_row_nodes(nodes, node_id, row_id) do
    node = get_node(nodes, node_id)
    up_id = node_up(node)
    down_id = node_down(node)
    col_id = node_column(node)

    # Restore node to column
    col = get_node(nodes, col_id)
    down = get_node(nodes, down_id)
    up = get_node(nodes, up_id)

    update_node(nodes, col_id, set_size(col, node_size(col) + 1))
    update_node(nodes, down_id, set_up(down, node_id))
    update_node(nodes, up_id, set_down(up, node_id))

    uncover_row_nodes(nodes, node_left(node), row_id)
  end

  # Choose column with minimum size (Knuth's S heuristic)
  defp choose_column(nodes, root_id) do
    root = get_node(nodes, root_id)
    choose_column_scan(nodes, node_right(root), root_id, nil, :infinity)
  end

  defp choose_column_scan(_nodes, col_id, root_id, best, _best_size) when col_id == root_id do
    best
  end

  defp choose_column_scan(nodes, col_id, root_id, best, best_size) do
    col = get_node(nodes, col_id)
    size = node_size(col)

    if size < best_size do
      choose_column_scan(nodes, node_right(col), root_id, col_id, size)
    else
      choose_column_scan(nodes, node_right(col), root_id, best, best_size)
    end
  end

  # DLX Search - Algorithm X with Dancing Links
  defp dlx_search(nodes, root_id, k, solution_agent, counter) do
    Agent.update(counter, &(&1 + 1))

    root = get_node(nodes, root_id)

    # If matrix is empty, we found a solution
    if node_right(root) == root_id do
      true
    else
      # Choose column with minimum size
      case choose_column(nodes, root_id) do
        nil ->
          false

        col_id ->
          col = get_node(nodes, col_id)

          # If column has no rows, no solution possible
          if node_size(col) == 0 do
            false
          else
            # Cover this column
            cover_column(nodes, col_id)

            # Try each row in this column
            result = try_rows(nodes, node_down(col), col_id, root_id, k, solution_agent, counter)

            # Uncover column
            uncover_column(nodes, col_id)

            result
          end
      end
    end
  end

  defp try_rows(_nodes, row_id, col_id, _root_id, _k, _solution_agent, _counter)
       when row_id == col_id do
    false
  end

  defp try_rows(nodes, row_id, col_id, root_id, k, solution_agent, counter) do
    row_node = get_node(nodes, row_id)

    # Add row to partial solution
    Agent.update(solution_agent, fn solution ->
      List.replace_at(solution, k, node_row_id(row_node))
    end)

    # Cover all other columns in this row
    cover_row_columns(nodes, node_right(row_node), row_id)

    # Recurse
    if dlx_search(nodes, root_id, k + 1, solution_agent, counter) do
      true
    else
      # Backtrack: uncover all columns in this row
      uncover_row_columns(nodes, node_left(row_node), row_id)

      # Try next row
      try_rows(nodes, node_down(row_node), col_id, root_id, k, solution_agent, counter)
    end
  end

  defp cover_row_columns(_nodes, node_id, row_id) when node_id == row_id, do: :ok

  defp cover_row_columns(nodes, node_id, row_id) do
    node = get_node(nodes, node_id)
    cover_column(nodes, node_column(node))
    cover_row_columns(nodes, node_right(node), row_id)
  end

  defp uncover_row_columns(_nodes, node_id, row_id) when node_id == row_id, do: :ok

  defp uncover_row_columns(nodes, node_id, row_id) do
    node = get_node(nodes, node_id)
    uncover_column(nodes, node_column(node))
    uncover_row_columns(nodes, node_left(node), row_id)
  end

  # Cover all columns in clue rows before search
  defp cover_clue_rows(nodes, clue_rows) do
    Enum.each(clue_rows, fn {_row_id, first_node_id} ->
      cover_clue_row(nodes, first_node_id, first_node_id)
    end)
  end

  defp cover_clue_row(nodes, node_id, first_node_id) do
    node = get_node(nodes, node_id)
    cover_column(nodes, node_column(node))

    next_id = node_right(node)
    if next_id != first_node_id do
      cover_clue_row(nodes, next_id, first_node_id)
    end
  end

  # Build DLX matrix for Sudoku
  defp build_dlx_matrix(nodes, puzzle) do
    # Create root node (id 0)
    root_id = 0
    make_node(nodes, root_id, -1, -1)

    # Create 324 column headers for Sudoku constraints
    num_cols = 324
    col_base = 1

    # Create column headers
    Enum.each(0..(num_cols - 1), fn i ->
      col_id = col_base + i
      make_node(nodes, col_id, -1, i)

      # Set column as its own column header
      col = get_node(nodes, col_id)
      update_node(nodes, col_id, set_column(col, col_id))
    end)

    # Link column headers horizontally
    Enum.each(0..(num_cols - 1), fn i ->
      col_id = col_base + i
      left_id = if i == 0, do: root_id, else: col_base + i - 1
      right_id = if i == num_cols - 1, do: root_id, else: col_base + i + 1

      col = get_node(nodes, col_id)
      col = set_left(col, left_id)
      col = set_right(col, right_id)
      update_node(nodes, col_id, col)

      if i == 0 do
        root = get_node(nodes, root_id)
        update_node(nodes, root_id, set_right(root, col_id))
      end

      if i == num_cols - 1 do
        root = get_node(nodes, root_id)
        update_node(nodes, root_id, set_left(root, col_id))
      end
    end)

    # Build rows
    node_id_counter = col_base + num_cols
    {_final_node_id, _row_count, row_metadata, clue_rows} =
      Enum.reduce(0..8, {node_id_counter, 0, %{}, []}, fn r, acc1 ->
        Enum.reduce(0..8, acc1, fn c, {node_id, row_count, meta, clues} ->
          cell_val = Enum.at(Enum.at(puzzle, r), c)

          if cell_val != 0 do
            # Cell has a clue - create only one row for that value
            d = cell_val
            {new_node_id, new_meta, first_node_id} = create_row(nodes, node_id, row_count, r, c, d, col_base, meta)
            {new_node_id, row_count + 1, new_meta, [{row_count, first_node_id} | clues]}
          else
            # Cell is empty - create rows for all possible values 1-9
            {final_node_id, final_count, final_meta} =
              Enum.reduce(1..9, {node_id, row_count, meta}, fn d, {nid, rc, m} ->
                {new_nid, new_m, _first} = create_row(nodes, nid, rc, r, c, d, col_base, m)
                {new_nid, rc + 1, new_m}
              end)

            {final_node_id, final_count, final_meta, clues}
          end
        end)
      end)

    {root_id, row_metadata, Enum.reverse(clue_rows)}
  end

  defp create_row(nodes, node_id, row_id, r, c, d, col_base, row_metadata) do
    # Calculate constraint column indices
    cell_col = col_base + r * 9 + c
    row_col = col_base + 81 + r * 9 + (d - 1)
    col_col = col_base + 162 + c * 9 + (d - 1)
    box_col = col_base + 243 + (div(r, 3) * 3 + div(c, 3)) * 9 + (d - 1)

    # Create 4 nodes
    node1_id = node_id
    node2_id = node_id + 1
    node3_id = node_id + 2
    node4_id = node_id + 3

    make_node(nodes, node1_id, row_id, -1)
    make_node(nodes, node2_id, row_id, -1)
    make_node(nodes, node3_id, row_id, -1)
    make_node(nodes, node4_id, row_id, -1)

    # Link nodes horizontally in a circular list
    node1 = get_node(nodes, node1_id)
    node1 = set_right(node1, node2_id)
    node1 = set_left(node1, node4_id)
    update_node(nodes, node1_id, node1)

    node2 = get_node(nodes, node2_id)
    node2 = set_right(node2, node3_id)
    node2 = set_left(node2, node1_id)
    update_node(nodes, node2_id, node2)

    node3 = get_node(nodes, node3_id)
    node3 = set_right(node3, node4_id)
    node3 = set_left(node3, node2_id)
    update_node(nodes, node3_id, node3)

    node4 = get_node(nodes, node4_id)
    node4 = set_right(node4, node1_id)
    node4 = set_left(node4, node3_id)
    update_node(nodes, node4_id, node4)

    # Insert each node into its column
    insert_into_column(nodes, node1_id, cell_col)
    insert_into_column(nodes, node2_id, row_col)
    insert_into_column(nodes, node3_id, col_col)
    insert_into_column(nodes, node4_id, box_col)

    new_metadata = Map.put(row_metadata, row_id, {r, c, d})

    {node_id + 4, new_metadata, node1_id}
  end

  defp insert_into_column(nodes, node_id, col_id) do
    node = get_node(nodes, node_id)
    col = get_node(nodes, col_id)

    # Set column reference
    node = set_column(node, col_id)

    # Insert at top of column (before col, after col.up)
    col_up_id = node_up(col)
    node = set_up(node, col_up_id)
    node = set_down(node, col_id)
    update_node(nodes, node_id, node)

    up = get_node(nodes, col_up_id)
    update_node(nodes, col_up_id, set_down(up, node_id))

    col = set_up(col, node_id)
    col = set_size(col, node_size(col) + 1)
    update_node(nodes, col_id, col)
  end

  # Decode solution from row IDs to Sudoku grid
  defp decode_solution(solution, puzzle, row_metadata) do
    # Start with given clues
    grid =
      Enum.map(puzzle, fn row ->
        Enum.map(row, & &1)
      end)

    # Decode each solution row using metadata
    Enum.reduce(solution, grid, fn row_id, acc ->
      case Map.get(row_metadata, row_id) do
        {r, c, d} when row_id >= 0 ->
          List.update_at(acc, r, fn row ->
            List.update_at(row, c, fn _ -> d end)
          end)

        _ ->
          acc
      end
    end)
  end
end

System.argv() |> DLX.main()

File: Algorithms/DLX/OCaml/dlx.ml
(* OCaml Dancing Links (DLX) Sudoku Solver *)
(* Port of C DLX implementation using mutable record fields *)

(* Type definitions - mutable record fields allow circular references *)
type dlx_node = {
  mutable left: dlx_node;
  mutable right: dlx_node;
  mutable up: dlx_node;
  mutable down: dlx_node;
  mutable column: dlx_node;
  mutable size: int;
  row_id: int;
  col_id: int;
}

(* Global iteration counter *)
let dlx_iterations = ref 0

(* Create a dummy node to serve as root/column header *)
let rec make_node row_id col_id =
  let rec node = {
    left = node;
    right = node;
    up = node;
    down = node;
    column = node;
    size = 0;
    row_id = row_id;
    col_id = col_id;
  } in
  node

(* Cover a column in the DLX matrix *)
let cover_column col =
  (* Remove column header from the header list *)
  col.right.left <- col.left;
  col.left.right <- col.right;

  (* For each row in this column *)
  let rec cover_rows row_node =
    if row_node != col then begin
      (* For each node in this row *)
      let rec cover_row_nodes right_node =
        if right_node != row_node then begin
          (* Remove this node from its column *)
          right_node.down.up <- right_node.up;
          right_node.up.down <- right_node.down;
          right_node.column.size <- right_node.column.size - 1;
          cover_row_nodes right_node.right
        end
      in
      cover_row_nodes row_node.right;
      cover_rows row_node.down
    end
  in
  cover_rows col.down

(* Uncover a column (exact reverse of cover) *)
let uncover_column col =
  (* For each row in this column (in reverse order) *)
  let rec uncover_rows row_node =
    if row_node != col then begin
      (* For each node in this row (in reverse order) *)
      let rec uncover_row_nodes left_node =
        if left_node != row_node then begin
          (* Restore this node to its column *)
          left_node.column.size <- left_node.column.size + 1;
          left_node.down.up <- left_node;
          left_node.up.down <- left_node;
          uncover_row_nodes left_node.left
        end
      in
      uncover_row_nodes row_node.left;
      uncover_rows row_node.up
    end
  in
  uncover_rows col.up;

  (* Restore column header to the header list *)
  col.right.left <- col;
  col.left.right <- col

(* Choose column with minimum size (Knuth's S heuristic) *)
let choose_column root =
  let min_col = ref None in
  let min_size = ref max_int in

  let rec scan col_node =
    if col_node != root then begin
      if col_node.size < !min_size then begin
        min_size := col_node.size;
        min_col := Some col_node
      end;
      scan col_node.right
    end
  in
  scan root.right;
  !min_col

(* DLX Search - Algorithm X with Dancing Links *)
let rec dlx_search root k solution =
  dlx_iterations := !dlx_iterations + 1;

  (* If matrix is empty, we found a solution *)
  if root.right == root then
    true
  else begin
    (* Choose column with minimum size *)
    match choose_column root with
    | None -> false
    | Some col ->
        (* If column has no rows, no solution possible *)
        if col.size = 0 then
          false
        else begin
          (* Cover this column *)
          cover_column col;

          (* Try each row in this column *)
          let rec try_rows row_node =
            if row_node == col then begin
              (* Uncover column and backtrack *)
              uncover_column col;
              false
            end else begin
              (* Add row to partial solution *)
              solution.(k) <- row_node.row_id;

              (* Cover all other columns in this row *)
              let rec cover_cols right_node =
                if right_node != row_node then begin
                  cover_column right_node.column;
                  cover_cols right_node.right
                end
              in
              cover_cols row_node.right;

              (* Recurse *)
              if dlx_search root (k + 1) solution then
                true
              else begin
                (* Backtrack: uncover all columns in this row *)
                let rec uncover_cols left_node =
                  if left_node != row_node then begin
                    uncover_column left_node.column;
                    uncover_cols left_node.left
                  end
                in
                uncover_cols row_node.left;

                try_rows row_node.down
              end
            end
          in
          try_rows col.down
        end
  end

(* Build DLX matrix for Sudoku *)
let build_dlx_matrix puzzle =
  (* Create root node *)
  let root = make_node (-1) (-1) in

  (* Create 324 column headers for Sudoku constraints *)
  (* 81 cell constraints + 81 row constraints + 81 col constraints + 81 box constraints *)
  let num_cols = 324 in
  let columns = Array.init num_cols (fun i -> make_node (-1) i) in

  (* Link column headers horizontally *)
  for i = 0 to num_cols - 1 do
    let col = columns.(i) in
    col.left <- (if i = 0 then root else columns.(i - 1));
    col.right <- (if i = num_cols - 1 then root else columns.(i + 1));
    if i = 0 then root.right <- col;
    if i = num_cols - 1 then root.left <- col;

    (* Set column as its own column header *)
    col.column <- col
  done;

  (* Store row metadata: map row_id to (r, c, digit) *)
  let row_metadata = Array.make 729 (0, 0, 0) in
  let row_first_nodes = Array.make 729 None in

  (* For each possible placement (row, col, digit) *)
  let row_count = ref 0 in
  for r = 0 to 8 do
    for c = 0 to 8 do
      if puzzle.(r).(c) <> 0 then begin
        (* Cell has a clue - create only one row for that value *)
        let d = puzzle.(r).(c) in
        let row_id = !row_count in
        row_metadata.(row_id) <- (r, c, d);

        (* Create 4 nodes for this placement (one per constraint type) *)
        let cell_col = r * 9 + c in
        let row_col = 81 + r * 9 + (d - 1) in
        let col_col = 162 + c * 9 + (d - 1) in
        let box_col = 243 + ((r / 3) * 3 + (c / 3)) * 9 + (d - 1) in

        let node1 = make_node row_id cell_col in
        let node2 = make_node row_id row_col in
        let node3 = make_node row_id col_col in
        let node4 = make_node row_id box_col in

        row_first_nodes.(row_id) <- Some node1;

        (* Link nodes horizontally in a circular list *)
        node1.right <- node2;
        node2.right <- node3;
        node3.right <- node4;
        node4.right <- node1;
        node1.left <- node4;
        node2.left <- node1;
        node3.left <- node2;
        node4.left <- node3;

        (* Insert each node into its column *)
        let insert_into_column node col_idx =
          let col = columns.(col_idx) in
          node.column <- col;
          node.up <- col.up;
          node.down <- col;
          col.up.down <- node;
          col.up <- node;
          col.size <- col.size + 1
        in

        insert_into_column node1 cell_col;
        insert_into_column node2 row_col;
        insert_into_column node3 col_col;
        insert_into_column node4 box_col;

        row_count := !row_count + 1
      end else begin
        (* Cell is empty - create rows for all possible values 1-9 *)
        for d = 1 to 9 do
          let row_id = !row_count in
          row_metadata.(row_id) <- (r, c, d);

          (* Create 4 nodes for this placement (one per constraint type) *)
          let cell_col = r * 9 + c in
          let row_col = 81 + r * 9 + (d - 1) in
          let col_col = 162 + c * 9 + (d - 1) in
          let box_col = 243 + ((r / 3) * 3 + (c / 3)) * 9 + (d - 1) in

          let node1 = make_node row_id cell_col in
          let node2 = make_node row_id row_col in
          let node3 = make_node row_id col_col in
          let node4 = make_node row_id box_col in

          row_first_nodes.(row_id) <- Some node1;

          (* Link nodes horizontally in a circular list *)
          node1.right <- node2;
          node2.right <- node3;
          node3.right <- node4;
          node4.right <- node1;
          node1.left <- node4;
          node2.left <- node1;
          node3.left <- node2;
          node4.left <- node3;

          (* Insert each node into its column *)
          let insert_into_column node col_idx =
            let col = columns.(col_idx) in
            node.column <- col;
            node.up <- col.up;
            node.down <- col;
            col.up.down <- node;
            col.up <- node;
            col.size <- col.size + 1
          in

          insert_into_column node1 cell_col;
          insert_into_column node2 row_col;
          insert_into_column node3 col_col;
          insert_into_column node4 box_col;

          row_count := !row_count + 1
        done
      end
    done
  done;

  (root, !row_count, row_metadata, row_first_nodes)

(* Decode solution from row IDs to Sudoku grid *)
let decode_solution solution num_rows puzzle row_metadata =
  let grid = Array.make_matrix 9 9 0 in

  (* Copy given clues *)
  for r = 0 to 8 do
    for c = 0 to 8 do
      if puzzle.(r).(c) <> 0 then
        grid.(r).(c) <- puzzle.(r).(c)
    done
  done;

  (* Decode each solution row using metadata *)
  for i = 0 to num_rows - 1 do
    let row_id = solution.(i) in
    if row_id >= 0 && row_id < Array.length row_metadata then begin
      let (r, c, d) = row_metadata.(row_id) in
      grid.(r).(c) <- d
    end
  done;

  grid

(* Cover given clues before starting search *)
let cover_clues puzzle row_metadata row_first_nodes =
  for r = 0 to 8 do
    for c = 0 to 8 do
      if puzzle.(r).(c) <> 0 then begin
        let d = puzzle.(r).(c) in
        (* Find the row for this clue *)
        let found = ref false in
        for row_id = 0 to Array.length row_metadata - 1 do
          if not !found then begin
            let (mr, mc, md) = row_metadata.(row_id) in
            match row_first_nodes.(row_id) with
            | Some first_node when mr = r && mc = c && md = d ->
                (* Cover all columns in this row *)
                let rec cover_row_cols node =
                  cover_column node.column;
                  if node.right != first_node then
                    cover_row_cols node.right
                in
                cover_row_cols first_node;
                found := true
            | _ -> ()
          end
        done
      end
    done
  done

(* Print puzzle *)
let print_puzzle grid =
  Printf.printf "\nPuzzle:\n";
  for r = 0 to 8 do
    for c = 0 to 8 do
      Printf.printf "%d " grid.(r).(c)
    done;
    Printf.printf "\n"
  done

(* Print raw board *)
let print_raw_board grid =
  for r = 0 to 8 do
    for c = 0 to 8 do
      Printf.printf "%d " grid.(r).(c)
    done;
    Printf.printf "\n"
  done

(* Read matrix file *)
let read_matrix_file filename =
  try
    Printf.printf "%s\n" filename;

    let ic = open_in filename in
    let puzzle = Array.make_matrix 9 9 0 in
    let row = ref 0 in

    try
      while !row < 9 do
        let line = input_line ic in
        let trimmed = String.trim line in
        if String.length trimmed > 0 && trimmed.[0] <> '#' then begin
          let parts = String.split_on_char ' ' trimmed
                     |> List.filter (fun s -> s <> "") in
          if List.length parts = 9 then begin
            List.iteri (fun col s ->
              puzzle.(!row).(col) <- int_of_string s
            ) parts;
            row := !row + 1
          end
        end
      done;
      close_in ic;
      Some puzzle
    with End_of_file ->
      close_in ic;
      if !row = 9 then Some puzzle else None
  with e ->
    Printf.fprintf stderr "Error reading file %s: %s\n" filename (Printexc.to_string e);
    None

(* Main solver *)
let solve_puzzle filename =
  match read_matrix_file filename with
  | None -> ()
  | Some puzzle ->
      (* Print initial board *)
      print_raw_board puzzle;
      print_puzzle puzzle;

      (* Build DLX matrix *)
      let (root, max_rows, row_metadata, row_first_nodes) = build_dlx_matrix puzzle in

      (* Cover clues before solving *)
      cover_clues puzzle row_metadata row_first_nodes;

      (* Solve *)
      dlx_iterations := 0;
      let solution = Array.make 81 (-1) in

      if dlx_search root 0 solution then begin
        (* Decode and print solution *)
        let solved = decode_solution solution 81 puzzle row_metadata in
        print_puzzle solved;
        Printf.printf "\nSolved in Iterations=%d\n\n" !dlx_iterations
      end else begin
        Printf.printf "No solution found\n"
      end

(* Main *)
let () =
  let start_time = Unix.gettimeofday () in

  (* Process each .matrix file from command line *)
  for i = 1 to Array.length Sys.argv - 1 do
    let filename = Sys.argv.(i) in
    if String.length filename >= 7 &&
       String.sub filename (String.length filename - 7) 7 = ".matrix" then
      solve_puzzle filename
  done;

  let end_time = Unix.gettimeofday () in
  Printf.printf "Seconds to process %.3f\n" (end_time -. start_time)
